quickfixj-core/src/main/java/quickfix/DefaultDataDictionaryProvider.java

```diff
@@ -20,54 +20,51 @@
 package quickfix;
 
 import org.quickfixj.QFJException;
+import org.quickfixj.SimpleCache;
 import quickfix.field.ApplVerID;
 
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-
 import static quickfix.MessageUtils.toBeginString;
 
 public class DefaultDataDictionaryProvider implements DataDictionaryProvider {
-    private final Map<String, DataDictionary> transportDictionaries = new ConcurrentHashMap<>();
-    private final Map<AppVersionKey, DataDictionary> applicationDictionaries = new ConcurrentHashMap<>();
-    private final boolean findDataDictionaries;
+    private final SimpleCache<String, DataDictionary> transportDictionaries;
+    private final SimpleCache<AppVersionKey, DataDictionary> applicationDictionaries;
 
     public DefaultDataDictionaryProvider() {
-        findDataDictionaries = true;
+        this(true);
     }
 
     public DefaultDataDictionaryProvider(boolean findDataDictionaries) {
-        this.findDataDictionaries = findDataDictionaries;
+        transportDictionaries = new SimpleCache<>(beginString -> {
+            if (findDataDictionaries) {
+                final String path = beginString.replace(".", "") + ".xml";
+                try {
+                    return new DataDictionary(path);
+                } catch (ConfigError e) {
+                    throw new QFJException(e);
+                }
+            }
+            return null;
+        });
+        applicationDictionaries = new SimpleCache<>(appVersionKey -> {
+            if (findDataDictionaries) {
+                final String beginString = toBeginString(appVersionKey.applVerID);
```

Comment:
I'm thinking to change this type of key from `AppVersionKey` to `ApplVerID` no instantiation will be needed.